packages/iTwinUI-react/src/core/Tree/Tree.tsx

```diff
@@ -224,40 +244,141 @@ export const Tree = <T,>(props: TreeProps<T>) => {
     return [flatList, firstLevelNodes];
   }, [data, getNode]);
 
-  return (
-    <ul
-      className={cx('iui-tree', className)}
-      role='tree'
-      onKeyDown={handleKeyDown}
-      ref={treeRef}
-      tabIndex={0}
-      onFocus={() => {
-        const items = getFocusableNodes();
-        if (items.length > 0) {
-          items[focusedIndex.current]?.focus();
-        }
-      }}
-      {...rest}
-    >
-      {flatNodesList.map((flatNode) => (
+  const itemRenderer = React.useCallback(
+    (index: number) => {
+      const node = flatNodesList[index];
+      return (
         <TreeContext.Provider
-          key={flatNode.nodeProps.nodeId}
+          key={node.nodeProps.nodeId}
           value={{
-            nodeDepth: flatNode.depth,
-            subNodeIds: flatNode.subNodeIds,
+            nodeDepth: node.depth,
+            subNodeIds: node.subNodeIds,
             groupSize:
-              flatNode.depth === 0
+              node.depth === 0
                 ? firstLevelNodesList.length
-                : flatNode.parentNode?.subNodeIds?.length ?? 0,
-            indexInGroup: flatNode.indexInGroup,
-            parentNodeId: flatNode.parentNode?.nodeProps.nodeId,
+                : node.parentNode?.subNodeIds?.length ?? 0,
+            indexInGroup: node.indexInGroup,
+            parentNodeId: node.parentNode?.nodeProps.nodeId,
+            scrollToParent: node.parentNode
+              ? () => {
+                  const parentNodeId = node.parentNode?.nodeProps.nodeId;
+                  const parentNodeIndex = flatNodesList.findIndex(
+                    (n) => n.nodeProps.nodeId === parentNodeId,
+                  );
+                  setScrollToIndex(parentNodeIndex);
+                }
+              : undefined,
           }}
         >
-          {nodeRenderer(flatNode.nodeProps)}
+          {nodeRenderer(node.nodeProps)}
         </TreeContext.Provider>
-      ))}
-    </ul>
+      );
+    },
+    [firstLevelNodesList.length, flatNodesList, nodeRenderer],
+  );
+
+  const [scrollToIndex, setScrollToIndex] = React.useState<number>();
+  const flatNodesListRef = React.useRef<FlatNode<T>[]>(flatNodesList);
+  flatNodesListRef.current = flatNodesList;
+
+  React.useEffect(() => {
+    setTimeout(() => {
+      if (scrollToIndex !== undefined) {
+        const nodeId = flatNodesListRef.current[scrollToIndex].nodeProps.nodeId;
+        const nodeElement = treeRef.current?.ownerDocument.querySelector(
+          `#${nodeId}`,
+        ) as HTMLElement;
+        nodeElement?.focus();
+        // Need to reset that if navigating with mouse and keyboard,
+        // e.g. going to the previous node with keyboard
+        // and then clicking the next node with mouse.
+        setScrollToIndex(undefined);
+      }
+    });
+  }, [scrollToIndex]);
+
+  const { outerProps, innerProps, visibleChildren } = useVirtualization({
```

Comment:
No need to initiate all the virtualization logic if it is not even enabled. Could you create a separate component with virtualization hook in there? Same as its done in `Combobox` for example

Reply:
Moved to separate component